### PR TITLE
New Adapter: Aso

### PR DESCRIPTION
**Pull request by adserver-online**
Aso server adapter added.

Docs PR: https://github.com/prebid/prebid.github.io/pull/5194

### CODE REVIEW

File: adapters/aso/aso.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	endpointParams := macros.EndpointTemplateParams{ZoneID: strconv.Itoa(params.Zone)}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func getMediaType(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	if bid.Ext != nil {
+		var bidExt openrtb_ext.ExtBid
+		err := json.Unmarshal(bid.Ext, &bidExt)
+		if err == nil && bidExt.Prebid != nil {
+			return openrtb_ext.ParseBidType(string(bidExt.Prebid.Type))
```
[the remark this answers is not in the document]
Comment (adserver-online) [reply adapters/aso/aso.go:143]: Mtype is part of OpenRTB 2.6, which our backend has not yet supported. We pass it explicitly in the adapter response but as a string in ext.prebid.type.
So, the suggestion is irrelevant.

File: static/bidder-info/bidgency.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+aliasOf: aso
+endpoint: "https://srv.bidgx.com/pbs/bidder?zid={{.ZoneID}}"
+maintainer:
+  email: "aso@bidgency.com"
```
Comment (gargcreation1992) [inline static/bidder-info/bidgency.yaml:4]: sent email for verification. Please reply with received.

Comment (adserver-online) [reply static/bidder-info/bidgency.yaml:4]: They replied, please check your inbox.

File: static/bidder-info/bcmint.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+aliasOf: aso
+endpoint: "https://srv.datacygnal.io/pbs/bidder?zid={{.ZoneID}}"
+maintainer:
+  email: "contact@bcm.ltd"
```
Comment (gargcreation1992) [inline static/bidder-info/bcmint.yaml:4]: Sent email to this address as well.

File: static/bidder-info/aso.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+endpoint: "https://srv.aso1.net/pbs/bidder?zid={{.ZoneID}}"
+maintainer:
+  email: "support@adsrv.org"
```
Comment (gargcreation1992) [inline static/bidder-info/aso.yaml:3]: Sent email to this address as well.

File: adapters/aso/aso.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+			errors = append(errors, &errortypes.BadInput{
+				Message: fmt.Sprintf("invalid bidderExt.Bidder, %s", err.Error()),
+			})
+			continue
+		}
+
+		requestCopy.Imp = []openrtb2.Imp{imp}
+
+		endpoint, err := a.buildEndpointURL(&impExt)
+
```
Comment (gargcreation1992) [inline adapters/aso/aso.go:62]: nitpick - please remove this extra blank line to be consistent with the rest of code.

File: adapters/aso/aso.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
```
Comment (gargcreation1992) [inline adapters/aso/aso.go:99]: you could also use built in - 
```
if adapters.IsResponseStatusCodeNoContent(responseData) {
		return nil, nil
}

if err := adapters.CheckResponseStatusCodeForErrors(responseData); err != nil {
       return nil, []error{err}
}
```

File: adapters/aso/aso.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+			resolveMacros(&bid)
+
+			bidType, err := getMediaType(bid)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+
+			b := &adapters.TypedBid{
+				Bid:     &bid,
```
Comment (gargcreation1992) [inline adapters/aso/aso.go:122]: Requesting you to please add the test case for multiple impressions returning multiple bids. 
I think this should be corrected as -
``` 
b := &adapters.TypedBid { 
    Bid: seatBid.Bid[i],
    //rest of code
}
```

File: adapters/aso/aso.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+
+	var errors []error
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bid := bid
```
Comment (gargcreation1992) [inline adapters/aso/aso.go:112]: this should be removed and instead use `seatBid.Bid[i]` at places where you would like to access bid.

Comment (adserver-online) [reply adapters/aso/aso.go]: Refactored.

File: adapters/aso/aso.go (hunk 6 above)
Comment (adserver-online) [reply adapters/aso/aso.go]: Refactored.

File: adapters/aso/aso.go (hunk 7 above)
Comment (adserver-online) [reply adapters/aso/aso.go]: Multi impressions/bids test added

File: static/bidder-info/bcmint.yaml (hunk 3 above)
Comment (adserver-online) [reply static/bidder-info/bcmint.yaml:4]: They replied, please check your inbox.

**Review by gargcreation1992 — APPROVED** [review]

**Review by Sonali-More-Xandr — APPROVED** [review]

**Pull request merged**